challenge-17/challenge-17.js

```diff
@@ -1,93 +1,120 @@
+			'abril' : '04',
+			'maio' : '05',
+			'junho' : '06',
+			'julho' : '07',
+			'agosto' : '08',
+			'setembro' : '09',
+			'outubro': '10',
+			'novembro': '10',
+			'dezembro': '12'
+		}
+		return moths[ monthName ];
+	}
 
-/*
-Agora iremos substituir as datas no formato "13 de junho de 1804" para
-"13/06/1804". A primeira coisa a fazer é criar uma função chamada
-`getMonthNumber`, que irá receber um parâmetro (será o nome do mês) e deverá
-retornar o número correspondente a esse mês.
-Ex: Se o usuário entrar com "março", deve retornar "03" (em string mesmo).
-Números com menos de dois dígitos devem ter um zero na frente.
-Crie então a função e mostre no console os retornos para os meses de março,
-setembro e dezembro.
-Use um console.log para cada mês, usando a frase:
-"O mês de [NOME DO MÊS] é representado pelo número [NÚMERO DO MÊS]."
-*/
-console.log( '\nMeses representados por números:' );
-// ?
+	console.log(  );
+	console.log( 'O mês de março é representado pelo número ' + getMonthNumber('março') + '.' );
+	console.log( 'O mês de setembro é representado pelo número ' + getMonthNumber('setembro') + '.' );
+	console.log( 'O mês de dezembro é representado pelo número ' + getMonthNumber('dezembro') + '.' );
+	/*
+	Agora, declare uma variável chamada `regexDate` que irá receber a expressão
+	regular que irá fazer o match com as datas. Crie grupos de captura para o
+	dia, o mês e o ano. Para os meses, você pode fazer o match somente com os
+	meses que estão no texto, não precisa adicionar todos.
+	Com o que vimos até agora, você consegue fazer :D
+	Mostre a regex no console.
+	*/
+	console.log( '\nRegex que vai fazer o match com as datas do texto:' );
+	var regexDate = / (\d\d) de (junho|julho) de (\d\d\d\d) /g;
 
-/*
-Agora, declare uma variável chamada `regexDate` que irá receber a expressão
-regular que irá fazer o match com as datas. Crie grupos de captura para o
-dia, o mês e o ano. Para os meses, você pode fazer o match somente com os
-meses que estão no texto, não precisa adicionar todos.
-Com o que vimos até agora, você consegue fazer :D
-Mostre a regex no console.
-*/
-console.log( '\nRegex que vai fazer o match com as datas do texto:' );
-// ?
-
-/*
-Agora crie a função que irá fazer o replace dos dados. A função será chamada
-de `replaceDate`. Ela deve retornar a data no formato:
-"[DIA]/[MÊS]/[ANO]"
-Após criar a função, faça o replace das datas no texto, mostrando no
-console o resultado.
-*/
-console.log( '\nReplace de datas:' );
-// ?
\ No newline at end of file
+	/*
+	Agora crie a função que irá fazer o replace dos dados. A função será chamada
+	de `replaceDate`. Ela deve retornar a data no formato:
+	"[DIA]/[MÊS]/[ANO]"
+	Após criar a função, faça o replace das datas no texto, mostrando no
+	console o resultado.
+	*/
+	console.log( '\nReplace de datas:' );
+	function replaceDate( regex, day, month, year ){
+		return day + '/' + getMonthNumber(month) + '/' + year ;
+	}
+	console.log( text.replace( regexDate, replaceDate ) );
```

Comment:
@andersonweb, funciona da seguinte forma:

você sabe que, funções em Javascript são objetos de primeira classe. Ou seja: você pode usá-las exatamente como usa qualquer objeto:
- criar funções literais;
- atribuir à variáveis;
- passar funções por parâmetro;
- serem retorno de outras funções;
- serem métodos de objeto;
etc.

Mas as funções tem uma funcionalidade especial: elas podem ser invocadas! Isso nós vimos nas primeiras aulas! Para invocar uma função, nós usamos os operadores `()`, correto?

Nas aulas de JS Funcional, nós vimos que podemos passar funções como parâmetro para outras funções, na forma de **referência**. Ou seja: você cria uma função externa, e passa só o nome dela como parâmetro. Quando a outra função for executada, ela irá invocar essa função que criamos.

Faça o seguinte teste para entender o que eu estou dizendo:

Você tem o código:

```js
function replaceDate( regex, day, month, year ) {
  return day + '/' + getMonthNumber(month) + '/' + year;
}
```

O que essa função espera, quando for invocada? Que você passa 4 parâmetros para ela: `regex`, `day`, `month` e `year`.

O primeiro parâmetro `regex`, nós nem estamos usando aqui, mas sabemos que o método `replace` retorna ele sempre como primeiro parâmetro da função de callback, certo?

Agora, ignore o método `replace` por um momento. Como você usaria essa função `replaceDate`?

Você usaria ela mais ou menos assim:

```js
console.log( replaceDate( 'uma regex qualquer', 10, 'agosto', 1990 ) );
```

E, por você estar **invocando** a função nesse momento, e passando os parâmetros pra ela, ela vai te retornar um resultado, que será:

```
10/08/1990
```

Certo? Nós nem usamos o método `replace` aqui, só invocamos a função de forma normal.

Quando você usa o `replace`, é como se você fizesse o seguinte:

```js
text.replace( regexDate, function replaceDate( regex, day, month, year ) {
  return day + '/' + getMonthNumber(month) + '/' + year;
} );
`` 

Agora olha esse código:

```js
console.log( text.replace( regexDate, replaceDate ) );
```

Qual a diferença de um para o outro? No de cima, você cria a função dentro do `replace`, mas não invoca. 

Na outra, a função `replaceDate` está **extraída**. Você pegou ela, colocou fora do `replace`, e passou uma **referência** a ela para o `replace`, sem invocar! A hora que o `replace` for feito, o próprio método se encarrega de invocar a função, já passando os parâmetros corretos.

Ficou mais claro?